File: app/src/main/java/com/tosh/poolandroid/LoginRegistration/RegisterActivity.java
package com.tosh.poolandroid.LoginRegistration;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.content.SharedPreferences;
import android.location.Location;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.text.TextUtils;
import android.view.View;
import android.widget.Toast;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputEditText;
import com.tosh.poolandroid.R;
import com.tosh.poolandroid.Retrofit.AuthRetrofitClient;
import com.tosh.poolandroid.Retrofit.NodeAuthService;

import java.util.regex.Pattern;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Retrofit;

import static androidx.core.util.PatternsCompat.EMAIL_ADDRESS;

public class RegisterActivity extends AppCompatActivity {

    private MaterialButton buttonLogin;
    private MaterialButton buttonRegister;
    private TextInputEditText inputName, inputEmail, inputPassword, inputConfirmPassword;

    private NodeAuthService api;
    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    private SharedPreferences preferences;
    private SharedPreferences.Editor editor;





    @Override
    protected void onStop() {
        compositeDisposable.clear();
        super.onStop();
    }

    @Override
    protected void onDestroy() {
        compositeDisposable.clear();
        super.onDestroy();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register);

        //init api
        Retrofit retrofit = AuthRetrofitClient.getInstance();
        api = retrofit.create(NodeAuthService.class);

        // views
        buttonLogin = findViewById(R.id.register_login_btn);
        buttonRegister = findViewById(R.id.register_btn);
        inputName = findViewById(R.id.name_register);
        inputEmail = findViewById(R.id.email_register);
        inputPassword = findViewById(R.id.password_register);
        inputConfirmPassword = findViewById(R.id.confirm_password_register);

        buttonLogin.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(RegisterActivity.this, LoginActivity.class);
                startActivity(intent);
                finish();
            }
        });

        buttonRegister.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                registerUser(inputName.getText().toString(),inputEmail.getText().toString(),
                        inputPassword.getText().toString(), inputConfirmPassword.getText().toString());
            }
        });
    }

    private void registerUser(final String name, final String email, String password, String confirmPassword) {
        if(TextUtils.isEmpty(name)){
            inputName.setError("Enter name");
            return;
        }
        if(TextUtils.isEmpty(email)){
            inputEmail.setError("Enter email");
            return;
        }

        if(!EMAIL_ADDRESS.matcher(email).matches()){
            inputEmail.setError("Enter valid email");
            return;
        }
        if(TextUtils.isEmpty(password)){
            inputPassword.setError("Enter password");
            return;
        }
        if(TextUtils.isEmpty(confirmPassword)){
            inputConfirmPassword.setError("Confirm password");
            return;
        }
        if(!TextUtils.equals(password, confirmPassword)){
            inputConfirmPassword.setError("Passwords don't match");
            return;
        }

        if(password.length()< 6){
            inputPassword.setError("Password should be atleast 6 characters");
        }

        compositeDisposable.add(api.registerUser(name,email,password,confirmPassword)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Consumer<String>() {
                    @Override
                    public void accept(String res) throws Exception {

                        if(res.equals("successful")){
                            addToSharedPreferences(email, name);
                            Intent intent = new Intent(RegisterActivity.this, PhoneActivity.class);
                            startActivity(intent);
                            finish();
                        }else{
                            Toast.makeText(RegisterActivity.this, ""+res, Toast.LENGTH_SHORT).show();
                        }
                    }
                }));
    }

    public void addToSharedPreferences(String email, String name){
        preferences = PreferenceManager.getDefaultSharedPreferences(this);
        editor = preferences.edit();
        editor.putString("email", email);
        editor.putString("name", name);
        editor.apply();

    }
}

File: app/src/main/java/com/tosh/poolandroid/view/activity/RegisterActivity.kt
package com.tosh.poolandroid.view.activity

import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.preference.PreferenceManager
import android.view.View
import android.view.WindowManager
import android.view.animation.DecelerateInterpolator
import android.view.animation.Interpolator
import android.widget.LinearLayout
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.util.PatternsCompat
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.agrawalsuneet.dotsloader.loaders.LazyLoader
import com.google.android.material.snackbar.Snackbar
import com.tosh.poolandroid.R
import com.tosh.poolandroid.viewmodel.MainViewModel
import kotlinx.android.synthetic.main.activity_register.*

class RegisterActivity : AppCompatActivity() {

    lateinit var containerLL: LinearLayout

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register)

        containerLL = findViewById<LinearLayout>(R.id.containerLL)
        this.getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN)

        register_login_btn.setOnClickListener {
            val intent = Intent(applicationContext, LoginActivity::class.java)

            startActivity(intent)
            finish()
        }

        register_btn.setOnClickListener {
            val name = name_register.text.toString().trim()
            val email = email_register.text.toString().trim()
            val password = password_register.text.toString().trim()
            val confirmPassword = confirm_password_register.text.toString().trim()

            if (name.isEmpty()){
                name_register.error = "Enter name"
                return@setOnClickListener
            }
            if (email.isEmpty()){
                email_register.error = "Enter email"
                return@setOnClickListener
            }
            if (!PatternsCompat.EMAIL_ADDRESS.matcher(email).matches()) {
                email_register.error = "Enter a valid email"
                return@setOnClickListener
            }
            if (password.isEmpty()){
                password_register.error = "Enter password"
                return@setOnClickListener
            }

            if (password.length <6){
                password_register.error = "Password should be more than 6 characters long"
                return@setOnClickListener
            }

            if (!password.equals(confirmPassword)){
                confirm_password_register.error = "Passwords dont match"
                return@setOnClickListener
            }
            loading()
            instantiateRegisterViewModel(name, email, password, confirmPassword)
        }
    }

    private fun instantiateRegisterViewModel(name:String, email:String,  password:String, confirmPassord:String) {
        containerLL.bringToFront()
        containerLL.visibility = View.VISIBLE
        val registrationViewModel: MainViewModel = ViewModelProviders.of(this)[MainViewModel::class.java]
        registrationViewModel.userRegister(name,email, password, confirmPassord).observe(this, Observer {

            if (it == "successful"){
                val intent = Intent(applicationContext, PhoneActivity::class.java)
                intent.putExtra("Name", name)
                intent.putExtra("Email", email)
                addToSharedPreferences(email)
                startActivity(intent)
                finish()
            }else{
                val contextView: View = findViewById(R.id.registerLayout)
                val snackbar = Snackbar
                    .make(contextView, it, Snackbar.LENGTH_LONG)
                snackbar.show()
                containerLL.visibility = View.GONE
            }
        })
    }

    fun addToSharedPreferences(email: String) {
        var preferences: SharedPreferences?
        var editor: SharedPreferences.Editor?

        preferences = PreferenceManager.getDefaultSharedPreferences(this)
        editor = preferences.edit()
        editor.putString("email", email)
        editor.apply()

    }

    private fun loading(){
        var lazyLoader = LazyLoader(this, 15, 5,
            ContextCompat.getColor(this, R.color.loader_selected),
            ContextCompat.getColor(this, R.color.loader_selected),
            ContextCompat.getColor(this, R.color.loader_selected))
            .apply {
                animDuration = 500
                firstDelayDuration = 100
                secondDelayDuration = 200
                interpolator = DecelerateInterpolator() as Interpolator
            }
        containerLL.addView(lazyLoader)
    }
}
